### PR TITLE
Added support for Composite Adversarial Attacks [JATIC-I2-IBM]

### PR DESCRIPTION
**Pull request by twweeb**
# Description

This pull request adds the support of the Composite Adversarial Attack (CAA) proposed in [1]. It provides an example notebook for explaining how to launch CAA.

[1] Towards Compositional Adversarial Robustness: Generalizing Adversarial Training to Composite Semantic Perturbations (CVPR 2023). [[Paper]](https://arxiv.org/abs/2202.04235), [[Demo]](https://hsiung.cc/CARBEN/), and [[Official repo]](https://github.com/IBM/composite-adv).

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Unit Test
- [x] Notebook Example

**Test Configuration**:
- OS: Linux
- Python version: 3.8.18
- ART version or commit number: [818b50e99cbf659560d1a31a9b5283a94896e488](https://github.com/Trusted-AI/adversarial-robustness-toolbox/commit/818b50e99cbf659560d1a31a9b5283a94896e488)
- PyTorch version: 2.0.1+cu117
- Kornia version: 0.7.0

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @twweeb Thank you very much for your pull request and contributing the Composite Adversarial Attack to ART! Your code looks good and I have added a few comments on the integration with ART. Please let me know what you think.

File: notebooks/composite-adversarial-attack.ipynb [hunk 1]
```diff
@@ -0,0 +1,292 @@
+{
```
[the remark this answers is not in the document]
Comment (twweeb) [reply notebooks/composite-adversarial-attack.ipynb:1]: Updated!

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 2]
```diff
@@ -0,0 +1,486 @@
+# MIT License
```
[the remark this answers is not in the document]
Comment (twweeb) [reply art/attacks/evasion/composite_adversarial_attack.py:1]: Updated!

File: tests/attacks/test_composite_adversarial_attack.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    n_train = 100
+    n_test = 11
+    yield x_train_cifar10[:n_train], y_train_cifar10[:n_train], x_test_cifar10[:n_test], y_test_cifar10[:n_test]
+
+
+@pytest.mark.skip_framework(
+    "tensorflow1", "tensorflow2", "tensorflow2v1", "keras", "non_dl_frameworks", "mxnet", "kerastf"
+)
+def test_generate(art_warning, fix_get_cifar10_subset):
+    print("test_generate")
```
Comment (beat-buesser) [inline tests/attacks/test_composite_adversarial_attack.py:46]:
```suggestion
```

File: tests/attacks/test_composite_adversarial_attack.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import logging
+
+import numpy as np
+import pytest
+import torch
```
Comment (beat-buesser) [inline tests/attacks/test_composite_adversarial_attack.py:22]:
```suggestion
```

File: tests/attacks/test_composite_adversarial_attack.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
+# MIT License
```
Comment (beat-buesser) [inline tests/attacks/test_composite_adversarial_attack.py:1]: Please move this file to `tests/attacks/evasion/`.

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 6]
```diff
@@ -0,0 +1,669 @@
...
+
+
+class CompositeAdversarialAttackPyTorch(EvasionAttack):
+    """
+    Implementation of the composite adversarial attack on image classifiers in PyTorch. The attack is constructed by
+    adversarially perturbing the hue component of the inputs. It uses order scheduling to search for the attack sequence
+    and uses the iterative gradient sign method to optimize the perturbations in semantic space and Lp-ball (see
+    `FastGradientMethod` and `BasicIterativeMethod`).
+
+    Note that this attack is intended for only PyTorch image classifiers with RGB images in the range [0, 1] as inputs.
```
Comment (beat-buesser) [inline art/attacks/evasion/composite_adversarial_attack.py:56]: I think we need the vertical line to guarantee a new line in the generated documentation.
```suggestion
    | Note that this attack is intended for only PyTorch image classifiers with RGB images in the range [0, 1] as inputs.
```

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+            adv_data = self.attack_pool_base[attack_idx](data, attack_parameter)
+
+        return adv_data, attack_parameter
+
+    def caa_hue(
+        self, data: "torch.Tensor", hue: "torch.Tensor", labels: "torch.Tensor"
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        Compute the adversarial examples for hue component.
+        :param data: A tensor of a batch of original inputs to be attacked.
```
Comment (beat-buesser) [inline art/attacks/evasion/composite_adversarial_attack.py:419]: We need an empty line for the automated documentation generation.
```suggestion
        Compute the adversarial examples for hue component.

        :param data: A tensor of a batch of original inputs to be attacked.
```

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 8]
```diff
@@ -0,0 +1,669 @@
...
+        return self._comp_pgd(
+            data=sur_data, labels=labels, attack_idx=0, attack_parameter=hue, ori_is_attacked=self.is_attacked.clone()
+        )
+
+    def caa_saturation(
+        self, data: "torch.Tensor", saturation: "torch.Tensor", labels: "torch.Tensor"
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        Compute the adversarial examples for saturation component.
+        :param data: A tensor of a batch of original inputs to be attacked.
```
Comment (beat-buesser) [inline art/attacks/evasion/composite_adversarial_attack.py:438]: We need an empty line for the automated documentation generation.
```suggestion
        Compute the adversarial examples for saturation component.

        :param data: A tensor of a batch of original inputs to be attacked.
```

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 9]
```diff
@@ -0,0 +1,669 @@
...
+            attack_parameter=saturation,
+            ori_is_attacked=self.is_attacked.clone(),
+        )
+
+    def caa_rotation(
+        self, data: "torch.Tensor", theta: "torch.Tensor", labels: "torch.Tensor"
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        Compute the adversarial examples for rotation component.
+        :param data: A tensor of a batch of original inputs to be attacked.
```
Comment (beat-buesser) [inline art/attacks/evasion/composite_adversarial_attack.py:461]: We need an empty line for the automated documentation generation.
```suggestion
        Compute the adversarial examples for rotation component.

        :param data: A tensor of a batch of original inputs to be attacked.
```

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 10]
```diff
@@ -0,0 +1,669 @@
...
+        return self._comp_pgd(
+            data=sur_data, labels=labels, attack_idx=2, attack_parameter=theta, ori_is_attacked=self.is_attacked.clone()
+        )
+
+    def caa_brightness(
+        self, data: "torch.Tensor", brightness: "torch.Tensor", labels: "torch.Tensor"
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        Compute the adversarial examples for brightness component.
+        :param data: A tensor of a batch of original inputs to be attacked.
```
Comment (beat-buesser) [inline art/attacks/evasion/composite_adversarial_attack.py:480]: We need an empty line for the automated documentation generation.
```suggestion
        Compute the adversarial examples for brightness component.

        :param data: A tensor of a batch of original inputs to be attacked.
```

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 11]
```diff
@@ -0,0 +1,669 @@
...
+                self.curr_dsm = sinkhorn_normalization(torch.exp(self.curr_dsm + dsm_grad + dsm_noise).detach())
+                self.curr_seq = hungarian(self.curr_dsm.detach())
+                dsm_noise = (torch.randn_like(self.curr_dsm) + 1) * 2  # Escaping local optimum
+        else:
+            raise ValueError()
+
+    def caa_attack(self, images: "torch.Tensor", labels: "torch.Tensor") -> "torch.Tensor":
+        """
+        The main algorithm to generate the adversarial examples for composite adversarial attack.
+        :param images: A tensor of a batch of original inputs to be attacked.
```
Comment (beat-buesser) [inline art/attacks/evasion/composite_adversarial_attack.py:624]:
```suggestion
        The main algorithm to generate the adversarial examples for composite adversarial attack.

        :param images: A tensor of a batch of original inputs to be attacked.
```

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 12]
```diff
@@ -0,0 +1,669 @@
...
+            adv_data = adv_data + self.step_size_pool[5] * torch.sign(_grad)
+            eta = torch.clamp(adv_data - sur_data, min=self.eps_pool[5][0], max=self.eps_pool[5][1])
+            adv_data = torch.clamp(sur_data + eta, min=0.0, max=1.0).detach_().requires_grad_()
+
+        return adv_data, eta
+
+    def update_attack_order(self, images: "torch.Tensor", labels: "torch.Tensor", adv_val: List) -> None:
+        """
+        Update the specified attack ordering.
+        :param images: A tensor of a batch of original inputs to be attacked.
```
Comment (beat-buesser) [inline art/attacks/evasion/composite_adversarial_attack.py:560]:
```suggestion
        Update the specified attack ordering.

        :param images: A tensor of a batch of original inputs to be attacked.
```

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 13]
```diff
@@ -0,0 +1,669 @@
...
+            attack_parameter=contrast,
+            ori_is_attacked=self.is_attacked.clone(),
+        )
+
+    def caa_linf(
+        self, data: "torch.Tensor", eta: "torch.Tensor", labels: "torch.Tensor"
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        Compute the adversarial examples for L-infinity (PGD) component.
+        :param data: A tensor of a batch of original inputs to be attacked.
```
Comment (beat-buesser) [inline art/attacks/evasion/composite_adversarial_attack.py:526]:
```suggestion
        Compute the adversarial examples for L-infinity (PGD) component.

        :param data: A tensor of a batch of original inputs to be attacked.
```

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 14]
```diff
@@ -0,0 +1,669 @@
...
+        :param labels: A tensor of a batch of the original labels to be predicted.
+        :param eta: The perturbation in the L-infinity ball.
```
Comment (beat-buesser) [inline art/attacks/evasion/composite_adversarial_attack.py:528]:
```suggestion
        :param eta: The perturbation in the L-infinity ball.
        :param labels: A tensor of a batch of the original labels to be predicted.
```

File: art/attacks/evasion/composite_adversarial_attack.py [hunk 15]
```diff
@@ -0,0 +1,669 @@
...
+            attack_parameter=brightness,
+            ori_is_attacked=self.is_attacked.clone(),
+        )
+
+    def caa_contrast(
+        self, data: "torch.Tensor", contrast: "torch.Tensor", labels: "torch.Tensor"
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        Compute the adversarial examples for contrast component.
+        :param data: A tensor of a batch of original inputs to be attacked.
```
Comment (beat-buesser) [inline art/attacks/evasion/composite_adversarial_attack.py:503]:
```suggestion
        Compute the adversarial examples for contrast component.

        :param data: A tensor of a batch of original inputs to be attacked.
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @twweeb Thank you very much for the updates. I have added my final review and found only a few minor required changes.

**Comment by twweeb** [comment]
All comments are addressed.

**Comment by beat-buesser** [comment]
Hi @twweeb Thank you very much for your first contribution to ART!

**Pull request merged**